src/Compilers/CSharp/Test/Semantic/Semantics/UnsafeTests.cs

```diff
@@ -936,18 +2002,47 @@ public void UnsafeIteratorSignatures()
                 Diagnostic(ErrorCode.ERR_UnsafeIteratorArgType, "p"));
 
             var withUnsafeOnMembers = string.Format(template, "", "unsafe");
-            CreateCompilation(withUnsafeOnMembers, options: TestOptions.UnsafeReleaseDll).VerifyDiagnostics(
-                // (4,64): error CS1637: Iterators cannot have pointer type parameters
-                Diagnostic(ErrorCode.ERR_UnsafeIteratorArgType, "p"),
-                // (4,56): error CS1629: Unsafe code may not appear in iterators
-                Diagnostic(ErrorCode.ERR_IllegalInnerUnsafe, "Iterator")); //this is for putting "unsafe" on an iterator, not for the parameter type
+            CreateCompilation(withUnsafeOnMembers, options: TestOptions.UnsafeReleaseDll, parseOptions: TestOptions.Regular12).VerifyDiagnostics(
+                // (4,70): error CS1637: Iterators cannot have pointer type parameters
+                //     unsafe System.Collections.Generic.IEnumerable<int> Iterator(int* p)
+                Diagnostic(ErrorCode.ERR_UnsafeIteratorArgType, "p").WithLocation(4, 70),
+                // (4,56): error CS8652: The feature 'ref and unsafe in async and iterator methods' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //     unsafe System.Collections.Generic.IEnumerable<int> Iterator(int* p)
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "Iterator").WithArguments("ref and unsafe in async and iterator methods").WithLocation(4, 56)); //this is for putting "unsafe" on an iterator, not for the parameter type
+
+            var expectedDiagnostics = new[]
+            {
+                // (6,9): error CS9231: Cannot use 'yield return' in an 'unsafe' block
+                //         yield return 1;
+                Diagnostic(ErrorCode.ERR_BadYieldInUnsafe, "yield").WithLocation(6, 9),
+                // (4,70): error CS1637: Iterators cannot have pointer type parameters
```

Comment:
>error CS1637: Iterators cannot have pointer type parameters [](http://example.com/codeflow?start=27&length=59)

Do we know what is the rational behind this error? Since the iterator is `unsafe`, why do we care if parameter is a pointer?

Reply:
I think that's because all iterator parameters need to be hoisted and we do not support hoisting pointers.

Reply:
> I think that's because all iterator parameters need to be hoisted and we do not support hoisting pointers.

The following works, however:
```
public class C
{
    unsafe public void* M(void* p)
    {
        var d = () => p;
        return d();
    }
}
```

Reply:
I see, interesting. I will note in the speclet as possible future work to allow pointer arguments in iterator/async methods. However, note that it is not a very useful scenario since `yield return` and `await` cannot be used in unsafe contexts anyway. We could allow that too (only disallow `yield`/`await` inside `fixed`), there is already a note in the speclet for that, but there might be some caveats (the speclet describes one with `stackalloc`).

Reply:
Update: now that we've switched to `unsafe` affecting only the signature, not the body, this would be more interesting to have. But still leaving as future work in the speclet for now.